### ISSUE TITLE
Report total number of TF memory cells?

### ISSUE THREAD

**Issue by obo** [ufal/neuralmonkey#287]
The number of network parameters reported is very useful. On the other hand, it does not reflect memory consumption, because e.g. batch size plays a big role bug does not add any new trained parameters.

Would there be a way to find and report the number of all the cells in all TF tensors? Could we add that to the report of network parameters?

**Comment by jindrahelcl** [comment]
What use is this statistics for? It probably won't reflect the actual memory consumption

Consider this example:

```
a = [30, 40]
b = a
```

Would you expect to see 1200 or 2400?

There are many many tensors along the way from the inputs to the output, most of them are generated automatically and many of them are not stored in memory.

Also, what is your definition of a _cell_? I am assuming it's a number in a tensor, but I might be wrong.

Maybe you mean something else - could you specify precisely what do you suggest? Possibly with examples.

**Comment by obo** [comment]
I'd expect 1200 here, but I'm not sure.

I'm interested in the number of floats across all tensors allocated at a GPU, so perhaps even this number separarely for every device in the session, including individual parameter servers of distributed TensorFlow in the long term.


26. ledna 2017 2:02:58 SEČ, Jindra Helcl <notifications@github.com> napsal:
>What use is this statistics for? It probably won't reflect the actual
>memory consumption
>
>Consider this example:
>
>```
>a = [30, 40]
>b = a
>```
>
>Would you expect to see 1200 or 2400?
>
>There are many many tensors along the way from the inputs to the
>output, most of them are generated automatically and many of them are
>not stored in memory.
>
>Also, what is your definition of a _cell_? I am assuming it's a number
>in a tensor, but I might be wrong.
>
>Maybe you mean something else - could you specify precisely what do you
>suggest? Possibly with examples.


-- 
Ondrej Bojar (mailto:obo@cuni.cz / bojar@ufal.mff.cuni.cz)
http://www.cuni.cz/~obo

**Comment by jindrahelcl** [comment]
Again, I can't see any use of this statistics. What do you want to gain by knowing that number. Why don't you just look at the total memory consumption and divide it by 32 bits?

Consider another example, just for fun:

```
a = tf.random_uniform([30,40])
b = tf.random_uniform([30,40])
c = tf.tanh(a + b)
```
2400 or 4800?

I believe that depends on how the memory usage optimization in Tensorflow is done. I assume that if c would be used on more places, it would be stored in memory, but when it's used only once, there is no reason for it to be stored. It gets complicated when you realize that there's gonna be a backward pass - should c be stored in memory or not? Depends on how complicated is to compute the gradients without knowing the values from the forward pass..

**Comment by jindrahelcl** [comment]
closing, feel free to reopen
